pysph/cpy/translator.py

```diff
@@ -693,3 +694,141 @@ def __init__(self, detect_type=ocl_detect_type, known_types=None):
 
     def _get_self_type(self):
         return KnownType('GLOBAL_MEM %s*' % self._class_name)
+
+
+class CUDAConverter(CConverter):
+    def __init__(self, detect_type=ocl_detect_type, known_types=None):
+        super(CUDAConverter, self).__init__(detect_type, known_types)
+        self.function_address_space = 'WITHIN_KERNEL '
+        self._known.update((
+            'LID_0', 'LID_1', 'LID_2',
+            'GID_0', 'GID_1', 'GID_2',
+            'LDIM_0', 'LDIM_1', 'LDIM_2',
+            'GDIM_0', 'GDIM_1', 'GDIM_2'
+        ))
+        self._local_decl = None
+
+    def _get_self_type(self):
+        return KnownType('GLOBAL_MEM %s*' % self._class_name)
+
+    def _get_function_args(self, node):
+        node_args = node.args.args
+        if PY_VER == 2:
+            args = [x.id for x in node_args]
+        else:
+            args = [x.arg for x in node_args]
+        annotations = self._annotations.get(node.name)
+        call_args = {}
+        if annotations:
+            for arg in args:
+                call_args[arg] = annotations.get(arg, Undefined)
+        else:
+            defaults = [ast.literal_eval(x) for x in node.args.defaults]
+
+            # Fill up the call_args dict with the defaults.
+            for i in range(1, len(defaults) + 1):
+                call_args[args[-i]] = defaults[-i]
+
+            # Set the rest to Undefined.
+            for i in range(len(args) - len(defaults)):
+                call_args[args[i]] = Undefined
+
+            call_args.update(self._known_types)
+
+        if len(self._class_name) > 0:
+            call_args['self'] = self._get_self_type()
+
+        call_sig = []
+        for arg in args:
+            value = call_args[arg]
+            type = self._detect_type(arg, value)
+            if self._local_decl and 'LOCAL_MEM' in type:
+                type = 'int'
+                arg = 'size_%s' % arg
+            call_sig.append('{type} {arg}'.format(type=type, arg=arg))
+
+        return ', '.join(call_sig)
+
+    def _get_local_info(self, obj):
+        fname = obj.__name__
+        annotations = self._annotations[fname]
+        local_info = {}
+        for arg, kt in annotations.items():
+            if 'LOCAL_MEM' in kt.type:
+                local_info[arg] = kt.base_type
+        if local_info:
+            return local_info
+        return None
+
+    def parse_function(self, obj):
+        src = dedent(inspect.getsource(obj))
+        fname = obj.__name__
+        self._annotations[fname] = getattr(obj, '__annotations__', None)
+        self._local_decl = self._get_local_info(obj)
+        code = self.convert(src)
+        self._local_decl = None
+        self._annotations = {}
+        return code
+
+    def visit_FunctionDef(self, node):
+        assert node.args.vararg is None, \
```

Comment:
Again a lot of similar code that is cut pasted.  Any cut pasting makes it very hard to update things later.  I don't mind adding an empty method to the base class that is only used for CUDA but you need to think about reducing cut/pasting and try to avoid sending PRs with any of it. :) Also please measure coverage for just the translator alone and ensure that the translator has almost 100% coverage.